### DISCUSSION TITLE
[Chapter2.5-2.10] 프로그램이 실행되었지만, 뭐가 뭔지 하나도 모르겠다

### DISCUSSION THREAD

**Discussion by ddongule** [elegant-functional-2023/secrets-of-computer#7] [General]
감상평을 남겨주세요~!
due: 6/29 11시 전!

➕ 연수) 제목에 Chapter 3이라 되어 있어 Chapter 2.5-10으로 수정했습니다!

**Comment by ella-yschoi** [comment]
## 감상평

- 이번 장은 특히 비유를 통한 설명 방식이 돋보였던 것 같아요.
- 콜백함수, 동기/비동기 등 자바스크립트를 통해 공부한 개념을 이 책에서 볼 줄이야..!
- 더 원론적으로 이해할 수 있어 좋았습니다.

<br/>

## 나누고 싶은 이야기

### 2.5장

- 책보다 더 쉽게 풀어 '[콜백 함수는 불닭볶음면 공장에서 시작된다](https://devella.oopy.io/4fc9d738-0670-4514-8a6b-18375aad44ce)' 라는 글로 써봤어요.
- 책에서는 도넛 공장으로 비유를 들어 설명하는데, 저는 불닭볶음면 공장으로 비유하고 자바스크립트 코드로 바꾸어 써봤습니다.
- 추가로 프로미스의 역할과도 연결지어 보았어요.

### 2.6장

- 동기와 비동기 또한 직장 상사와의 업무 이야기로 비유를 들어 설명하는데 정말 좋은 설명이라 생각했습니다.
- 평소 콜백 함수, 프로미스, 비동기 동기 개념을 따로따로 이해하고 있어서 머릿속에 잘 정리되지 않는 느낌이었는데요.
- 책을 통해 다시 보니 이들은 꽤나 촘촘하게 엮여있다고 느끼기도 했습니다.

### 다음주에 뵈어용

- 엄마 모시고 🇯🇵 잘 다녀오겠습니당
- 즐겁고 유익한 스터디 시간 되시길!


<br/>

**Reply by ddongule** [reply]
ㄷㄷ 글 불닭 글 짱이네요 멋져요!!!!!!!!!!!!!!! 정리왕 연수님 👍
여행재밌게다녀오세요!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! 갸악

**Comment by ddongule** [comment]
## 감상평

코루틴이라는 개념을 이해하기 위해 콜백부터 빌드업하는 과정이 아주 인상깊었다. (갓샤오펑 쵝오) 
콜백함수, 동기/비동기, 블로킹/논블로킹 등 익숙한 개념들이 나와서 읽으면서 쬑굼 신났다. 
이걸 활용해서 코루틴을 한번 직접 작성해보고 싶다는 생각도 들었는데 아직 어떤 상황에 사용해야 잘 사용하는건지는 죄금 애매모호하다. 🤔

으하하 2장 정리 중에 있는데 속도가 안난다 큰일이다 ㄷㄷ ㅎㅎㅎㅎㅎㅎㅎㅎ ;; 우헤헤
저는 수목금 워크샵을 다녀와야해서 주말부터 짬짬이 읽어봤습니다.. 흑흑.. 
토요일에 이 내용이 기억이 날지 잘 모루겠다. 으하하.

**Reply by ella-yschoi** [reply]
오오... 코루틴 직접 작성해보기!! 재밌겠는데요!!
워크샵 길게 가시네용 잘 다녀오세요 밍경니임~~~!! ❣️

**Reply by sunhpark42** [reply]
async, await 를 사용하고 있는 것 부터 코루틴을 활용한 코드가 아닐까 싶어요잉

**Comment by yunseorim1116** [comment]
# 느낀점

블로킹,논블로킹 방식에 대한 개념을 한번 더 훑고 지나갈 수 있어서 좋았다. 이벤트 기반 프로프로그래밍은 자바스크립트에서 시작된 개념인줄로만 알았는데 ~~(이것도 모른다고 생각할수 있지만..정말 충격)~~ 이 책에 있는 설명을 보고 자바스크립트 이벤트루프에 대해서 좀 더 깊게 파보고싶어졌다.

- v8실행엔진에서의 이벤트 루프는 어떤 스레드에서 동작하고 (=메인스레드라고 한다.)
- 어떤 코드를 통해 동작하는지
- 왜 js는 이벤트 기반 프로그래밍을 사용하는지 구체적인 이유

가 궁금해졌다.  

*이벤트 루프는 자바스크립트 엔진이 아닌 **런타임**에서 가지고 있는 하나의 장치고, 단지 자바스크립트 코드를 실행하기위해 자바스크립트 엔진을 이용할 뿐이다. (역주: 실제로 V8 엔진에는 이벤트 루프를 관리하는 코드가 없다.)

```jsx
void MessagePumpDefault::Run(Delegate* delegate) {
  AutoReset<bool> auto_reset_keep_running(&keep_running_, true);

  for (;;) {
#if defined(OS_APPLE)
    mac::ScopedNSAutoreleasePool autorelease_pool;
#endif

    Delegate::NextWorkInfo next_work_info = delegate->DoWork();
    bool has_more_immediate_work = next_work_info.is_immediate();
    if (!keep_running_)
      break;

    if (has_more_immediate_work)
      continue;

    has_more_immediate_work = delegate->DoIdleWork();
    if (!keep_running_)
      break;

    if (has_more_immediate_work)
      continue;

    if (next_work_info.delayed_run_time.is_max()) {
      event_.Wait();
    } else {
      event_.TimedWait(next_work_info.remaining_delay());
    }
    // Since event_ is auto-reset, we don't need to do anything special here
    // other than service each delegate method.
  }
}
```

*실제 이벤트 루프가 어떻게 구현되어 있는지 알아본 사람의 블로그를 보고 코드를 가지고 왔다. (정말정말정말대박존경스럽....)

사실 작년에 봤을때 제대로 이해하지 못했는데.. 다시 보니 정말 새롭다. 언어,런타임,실행엔진 레이어에 대한 개념도 깊게 생각해본적 없었는데 책을 읽으면서 한번 더 고민해보게 되며 점점 개념이 확립이 되어가는 느낌이었다. 또한 이벤트루프,이벤트핸들러,핸들러 함수 등 어째서 네이밍이 이렇게 되었는지에 대해 나는 한번이라도 고민해본적이 있는가? 에 대한 생각이 들었다. 

# 궁금증

- i/o 작업은 정확히 무엇일까?
- db작업도 i/o작업에 포함이 되는 것일까 ?
- 동기적/비동기적 처리에 대한 차이점이 존재하는데 언어마다 설계가 다른건가 ? 그럼 동기적으로 작동하는 언어도 있는가..?
- 데이터베이스 스레드란 무엇인가
- 주 스레드가 사용자 요청을 처리하는 작업과 데이터 베이스 스레드가 데이터 베이스를 처리하는 작업을 동시에 진행할 수 있다. 이것이 바로 비동기가 가지는 높은 효율성이다.

⇒ 자바스크립트가 비동기 언어인 이유는 스레드가 하나이기 때문에 어쩔 수 없이 비동기로 설계된다고 생각했는데 그게 아니었던건가..? 그럼 단일 스레드인데 동기적으로 설계도 가능한가? 

- 알림작동방식이란 무엇일까?
- epoll
- 자바스크립트에서는 웹브라우저 런타임에서 메인 스레드가 하나기 때문에 스레드풀을 구현할수 있는 & 혹은  이미 구현되어있는 무언가가 있을까?
- 프로시저 호출, rpc,스케줄링
- 제너레이터나 코루틴처럼 실행을 미루는 방식을 커링을 사용해서 구현할수 있을까? (라는 불투명한 생각) 애초에 둘을 비교하는게 맞나..아닌거같고..

라는 궁금증들이 있었고 챗지피티를 사용해 답변을 얻어냈지만 구글링을 통한 크로스체킹을 하지 못한 관계로.. 좀더 공부가 필요할 것 같다는 생각이 든다. 
이번주는 월-금까지 매일 야근을 할정도로 회사일이 바빴는데 그 와중에 책 내용이 정말 너무 재밌고 유익해서 좋았다. 읽으면 읽을수록 이 책으로 스터디를 하길 잘한 것 같다는 생각이 들었다. 하지만 책 내용 기반으로 구현을 해보지 못하고 공부시간이 부족해서 계속 머릿속에 **‘??????????**’가 뜨는 이슈는 지울수없음……….아쉽………

어떤 이슈냐 하면,,

- 멀티 스레드를 사용하면 좋다⇒ 캔버스나 애니메이션같은 기술을 제대로 다루어본적이 없기때문일까 ? 사실 프론트 구현하면서 멀티스레드를 사용해야 할 상황을 체감해본적이 없음
- 멀티스레드는 보통 어떤 상황에 사용해야하는지, 기술을 사용할 때 무조건적으로 좋은건 없을텐데 프로덕트에서 멀티스레드를 구현할때 단점은 무엇인지 (코드의 복잡도가 올라간다던지) 그런것들이 제대로 와닿지 않음
- 노드로 워커스레드를 통해 멀티 스레드를 돌려보긴 했는데 포스트 메세지로 인자만 주고받는 느낌이어서 이게 그렇게 좋고 빠른건가? 정석으로 멀티 스레드를 사용해 본적이 없으니 **비교가 안되니까 좀 답답한 느낌**

**Comment by sa02045** [comment]
[정리](https://victorywin.notion.site/2-9cdcb578aa404da5a96a570e1942c206)

동기/비동기/블로킹/논블로킹 매번 헷갈리는 개념이었는데 책을 읽으니 어느정도 정리가 되어서 좋았습니다~
- 블로킹/논블로킹은 스레드, 운영체제를 일시정지하느냐에 대한 관점
- 노는 리소스없이 최대한 효율적으로 리소스를 활용하자! -> 동시성 -> 동시성을 어떻게 구현해야할까? 
   - 다중 프로세스: 매번 프로세스를 생성, 종료는 오버헤드. 통신도 어려움
   - 다중 스레드: 스레드 안전문제가 있긴 하지만 차선책임
- 리소스 최대한 활용을 위해 기다리지 않아도 되는 비동기로 처리하자(실행이 완료되면 나에게 알려줘)
- Node.js의 공식문서에 나온 설명도 같이 읽어보면 좋을 것 같아요  https://nodejs.org/en/about. `Node.js의 어떤 함수도 직접 입출력을 수행하지 않기 때문에 Node.js 표준 라이브러리의 동기식 메서드를 사용하여 입출력을 수행하는 경우를 제외하고는 프로세스가 절대 블록되지 않습니다 `
- 프로세스/스레드/코루틴은 서로 다른 레이어에서 실행흐름을 제어한다!

**Reply by sa02045** [reply]
제너레이터 사용 예시)

- Stream SSR 
- https://github.com/rotoshine/infcon2023-sample-streaming-ssr/blob/a1dddae83f9a7086500bc135a386330b33e6f9c2/src/entry-server.tsx#L27
- 출처) 인프콘 2023 
<img width="1423" alt="스크린샷 2024-06-29 오전 11 11 40" src="https://github.com/elegant-functional-2023/secrets-of-computer/assets/50866506/7fca08d6-08e8-45fd-bfd4-69c75ab1182e">

**Reply by sa02045** [reply]
- Next.js Stream SSR도 제너레이터로 구현되어있을까?

**Comment by iborymagic** [comment]
감상)
- 동기/비동기와 블로킹/논블로킹을 구분하는 건 여전히 헷갈려서 한 번 더 읽었읍니다.. 각각에 대한 개념은 명확한데, 둘을 비교하는 단계에 오니 처음에는 약간 헷갈렸읍니다..
- 반응자 패턴을 카페에 비유해주니 내용이 잘 들어왔읍니다.. 근데 이벤트 순환은 이벤트를 기다렸다가 핸들러를 호출하는 일련의 반복적인 과정을 모델화한 것을 말한다고 볼 수 있을 것 같은데, 그러면 얘는 고객보다는 카운터 직원에 더 가깝지 않나? 하는 생각을 했읍니다..
- 후반부는 막 되게 새로운 개념을 배웠다기보다는, 이전까지 배웠던 개념들을 약간 복습하는 느낌이었던 것 같읍니다..
- 책을 시간 날 때마다 틈틈이 끊어서 읽었더니 내용이 기억이 잘 안납니다.. 한 번에 주우욱 읽는 게 내용이 더 잘 들어오는 것 같읍니다..
- 작가분이 윤태진씨를 엄청 좋아하시는 것 같읍니다..

메모)
- 커널은 스레드를 보고 CPU 리소스를 할당, 코루틴은 보지 않음. 스레드 = 커널 상태 스레드
- 스레드에 할당된 시간 내에 실행할 코루틴을 사용자가 결정. 즉, CPU는 스레드에 시간을 할당하고 그걸 사용자가 사용자 상태에서 재차 코루틴에게 할당. 그래서 코루틴 = 사용자 상태 스레드

**Reply by ella-yschoi** [reply]
8쪽 밑에서 세 번째 단락을 보면 옮긴이가 윤태진 아나운서를 우상으로 칭하며 책에 이름을 써도 되냐는 부탁을 하셨었대요 ㅎㅎㅎ

**Comment by thgr8ganzi** [comment]
감상평

1. 보통 논블로킹, 블록킹을 공부하면 제어권이라는 단어가 나와서 아~ 이해했구나 하고 넘어갔는데 책을 읽고나서 조금 매해모호 했던 부분이 역시나 애매모호 해졌다.
2. 이벤트 기반 프로그래밍은 듣기만 했지 공부한적은 없었는데 프로듀서 컨슈머 패턴과 유사성을 느꼇다... 이벤트 기반 프로그래밍의 I/O 작업 하나씩 떼어놓으면 handeler 가 처리해서 컨슈머에게 제공하는것 같은 느낌..?
3. 코루틴은 어렵다... 더 공부 해야겠다.

**Reply by ella-yschoi** [reply]
'매해모호'했던 부분이 애매모호 해졌군요 (ㅋㅋㅋㅋ)

**Comment by sunhpark42** [comment]
밑줄친 부분들
- 동기 호출은 호출자와 수신자가 같은 스레드에서 실행 중인지 여부와는 관련이 없다는 결론을 내릴 수 있습니다. 
- 주 스레드의 ‘유휴 시간’이 없어진 대신 그 자리를 끊임없는 작업, 작업, 작업이 차지하고 있습니다. 또 데이터베이스 스레드에도 빈 자리가 거의 없으며 마찬가지로 작업, 작업, 작업이 자리를 차지하고 있습니다. 
- 모든 비동기 호출이 반드시 동기 호출보다 더 효율적인 것은 아니기 때문에 구체적인 상황에 따라 분석해야 합니다. 
- CPU 제어권을 상실했다가 되찾는 시간 동안 스레드나 프로세스는 블로킹되어 일시 중지됩니다. 
- 운영 체제는 CPU의 리소스를 최대한 활용할 수 있도록 각 스레드 간에 CPU 사용 시간을 효율적으로 할당해야 하는데, 이것이 바로 블로킹 입출력 방식이 필요한 핵심적인 이유입니다. 그 결과, 시간이 많이 걸리는 입출력 작업이 포함될 때 가끔 호출 스레드가 블로킹되며 일시 중지되는 일이 발생합니다. 

소감
- 운영체제의 핵심은 CPU를 어떻게 최대한, 효율적으로 사용할 수 있을까라는 것을 판단하는데 있다는 것을 다시 한 번 상기하게 됨.

**Reply by ella-yschoi** [reply]
'운영체제의 핵심은 CPU를 어떻게 최대한, 효율적으로 사용할 수 있을까라는 것을 판단하는데 있다는 것' 이 문장 공감합니당
저도 이 부분을 포인트 삼으면서 계속 운영체제 부분을 읽고 있어용

**Comment by sunhpark42** [comment]
한결님, 동희님, 주현님 다음주까지 정리..화이팅~

**Reply by hanana1253** [reply]
저는 CH2는 여기에 이어서 정리중..
https://teal-line-f3e.notion.site/CH-02-b12750e6ede7459aa12b353563e53d95?pvs=4
